File: past-competitions/2018/src/brett-cpp/Prob12.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int main(int argc, char *argv[])
{
	// open file for reading
	FILE *inputFile = fopen("Prob12.in.txt", "r");
	int testCases;

	// read number of test cases
	fscanf(inputFile, "%d\n", &testCases);
	
	int limit;
	int sieve[45000];
	int primes[45000];
	int primeCount = 0;
	int compositeCount = 0;

	// execute test cases
	while(testCases > 0){
		fscanf(inputFile, "%d\n", &limit);

		// reset the sieve
		// start with 2, as 0 and 1 are considered neither prime nor composite
		for(int i = 2; i <= limit; i++){
			// 1 == prime
			// a number will be set to 0 when found to be composite
			sieve[i] = 1;
		}
		// reset the number of primes
		// we don't have to reset the list of primes, since
		// which numbers are prime won't change
		primeCount = 0;

		// perform the sieve
		for(int i = 2; i <= limit; i++){
			if(sieve[i] == 1){
				// prime number
				// eliminate multiples as composites and count them
				compositeCount = 0;
				for(int j = i + i; j <= limit; j = j + i){
					if(sieve[j] == 1){
						// only count if not already eliminated
						sieve[j] = 0;
						compositeCount++;
					}
				}
				// if eliminations occurred, print how many
				if(compositeCount > 0){
					printf("Prime %d Composite Set Size: %d\n", i, compositeCount);
				}
				// record the prime
				primes[primeCount] = i;
				primeCount++;
			}
		}

		// print the primes
		printf("{");
		for(int i = 0; i < primeCount; i++){
			if(i > 0){
				// add commas as needed
				printf(",");
			}
			printf("%d", primes[i]);
		}
		printf("}\n");

		testCases = testCases - 1;
	}

	// close the file
	fclose(inputFile);
}

File: past-competitions/2018/src/brett-python/Prob12.py
# Open the input file
with open("Prob12.in.txt", "rt") as inputFile:
    # Read the number of test cases (trim out the newline)
    cases = int(inputFile.readline().replace("\n", ""))
    # For each test case
    for caseNum in range(cases):
        # Read the limit; add 1 so the limit itself is included
        limit = int(inputFile.readline().replace("\n", "")) + 1
        # Fill an array with 'True' to indicate numbers
        # are potential primes. We'll mark them as False
        # as they are proven to be composite
        possiblePrimes = [True for x in range(limit)]
        primes = []
        # start with 2; 0 and 1 are neither prime nor composite
        for i in range(2, limit):
            if possiblePrimes[i]:
                # if still True, it's a prime
                primes.append(i)
                eliminated = 0
                # eliminate multiples
                # go from 2i to limit, increasing by i
                for j in range(i * 2, limit, i):
                    if possiblePrimes[j]:
                        possiblePrimes[j] = False
                        eliminated = eliminated + 1
                # if we eliminated anything new, report it
                if eliminated > 0:
                    print("Prime " + str(i) + " Composite Set Size: " + str(eliminated))
        # When done, report all primes
        first = True
        # defining end suppresses the default newline
        print("{", end="")
        for prime in primes:
            if not first:
                # add commas as needed
                print(",", end="")
            first = False
            print(str(prime), end="")
        # now print the newline
        print("}")
